Adding L2L3, L3L4, L5L6 seeds into TrackletProcessor

jasonfan393 opened the pull request:
PR generalizes TrackletProcessor code to work in other barrel seeds besides L1L2.

pwittich commented:
@jasonfan393 there are some files that changed here that are not related to the tracklet processor (the gitlab-ci.yml file and the script file for the ME, for instance.) Are these changes intentional?

jasonfan393 commented:
> @jasonfan393 there are some files that changed here that are not related to the tracklet processor (the gitlab-ci.yml file and the script file for the ME, for instance.) Are these changes intentional?

No, they aren't, thanks.

pwittich commented:
@jasonfan393 for some reason the automated checks are failing; can you look into why? It appears to be related to the changes to `download.sh` that you made.

jasonfan393 commented:
> @jasonfan393 for some reason the automated checks are failing; can you look into why? It appears to be related to the changes to `download.sh` that you made.

One issue was just me including commits that have already been merged, which I just fixed. There seems now to be an issue with  kNTEUnits which I changed in Constants.h to an array. I hadn't realized this is also used in VMRouterCM, trying to see how I can fix this now.

aehart commented on TestBenches/TrackletProcessor_L1L2D_test.cpp:
```diff
@@ -8,7 +8,7 @@
 #include "Constants.h"
 
-const int nevents = 100;  //number of events to run
+const int nevents = 10;  //number of events to run
```
@jasonfan393 Is there a reason for reducing the number of events being tested here? Similar comments apply to `nevents` in the other TP test benches.

jasonfan393 replied:
Nope, I had just changed this while testing and commited it

aehart commented on TrackletAlgorithm/TrackletEngineUnit.h:
```diff
@@ -31,22 +31,52 @@ class TrackletEngineUnit {
-
-    ap_uint<1> stubptinnertmp[kNBitsPTLut] =
+    if (Seed==TF::L1L2&&iTC==3){
+      ap_uint<1> stubptinnertmp[256]=
```
@jasonfan393 Please remove the 256 specifying the size of the array. It's a magic number and isn't necessary when the array is being initialized at the same time.

This should also be an array of type `const ap_uint<1>`, and these comments apply to the other "tmp" arrays defined below.

aehart commented on TrackletAlgorithm/TrackletProcessor.h:
```diff
@@ -316,10 +360,10 @@ TC::barrelSeeding(const AllStub<InnerRegion> &innerStub, const AllStub<OuterRegi
 
-  const ap_int<TrackletParameters::kTParPhi0Size + 2> phicrit = *phi0 - (*rinv<<1);
-  const bool keep = (phicrit > 9253) && (phicrit < 56269);
+  const ap_int<TrackletParameters::kTParPhi0Size + 2> phicrit = *phi0 - (*rinv>>8)*469;//fixme, calculation adjusted to match emulation
+  const bool keep = (phicrit > 10190) && (phicrit < 55345);
```
@jasonfan393 Could you please eliminate the magic numbers (469, 10190, and 55345) by defining meaningfully named constants in Constants.h? I think you can follow the same naming scheme as is used in the emulation.

aehart commented on TrackletAlgorithm/TrackletProcessor.h:
```diff
@@ -360,8 +404,8 @@ TC::addProj(const TrackletProjection<TProjType> &proj, const BXType bx, Tracklet
     projout[1].write_mem(bx, proj, nproj[1]++);
-  if (NProjOut > 2 && TPROJMask & (0x1 << 2) && success && proj_success && phi == 2)
-    projout[2].write_mem(bx, proj, nproj[2]++);
+  if (NProjOut > 2 && TPROJMask & (0x1 << 2) && success && proj_success && phi == 2){
+    projout[2].write_mem(bx, proj, nproj[2]++);}
```
@jasonfan393 Please remove the curly braces.

aehart commented on TrackletAlgorithm/TrackletProcessor.h:
```diff
@@ -445,6 +488,19 @@ TC::processStubPair(
+        const TrackletProjection<BARREL2S> tproj_L4(TCID, trackletIndex, phiL[1], zL[1], der_phiL, der_zL);
+        const TrackletProjection<BARREL2S> tproj_L5(TCID, trackletIndex, phiL[2], zL[2], der_phiL, der_zL);
+        const TrackletProjection<BARREL2S> tproj_L6(TCID, trackletIndex, phiL[3], zL[3], der_phiL, der_zL);
+	TC::addProj<BARRELPS, nproj_L1, ((TPROJMaskBarrel & mask_L1) >> shift_L1)> (tproj_L1, bx, &projout_barrel_ps[L1PHIA], &nproj_barrel_ps[L1PHIA], success && valid_proj[0]);
+	addL4 = TC::addProj<BARREL2S, nproj_L4, ((TPROJMaskBarrel & mask_L4) >> shift_L4)> (tproj_L4, bx, &projout_barrel_2s[L4PHIA], &nproj_barrel_2s[L4PHIA], success && valid_proj[1]);
```
@jasonfan393 Please indent these lines to match the neighboring lines.

aehart commented on TrackletAlgorithm/TrackletProcessor.h:
```diff
@@ -635,26 +692,26 @@ TrackletProcessor(
-
-    const TrackletProjection<BARRELPS>::TProjTCID TCId(iTC);
+    //fix later
+    const TrackletProjection<BARRELPS>::TProjTCID TCId(iTC+Seed*16);
```
@jasonfan393 Please eliminate magic number 16 by replacing with
```c++
    const TrackletProjection<BARRELPS>::TProjTCID TCId(iTC + (Seed << TrackletProjection<BARRELPS>::kTProjITCSize));
```

aehart commented on project/script_TP.tcl:
```diff
@@ -11,24 +11,25 @@ open_project -reset trackletProcessor
 
 # Define FPGA, clock frequency & common HLS settings.
 source settings_hls.tcl
+set_clock_uncertainty .05
```
@jasonfan393 Is this really necessary? Given how unreliable the timing estimates from HLS are, this doesn't seem realistic.

jasonfan393 replied:
as of now, the L3L4 TP comes pretty close to the 4ns threshold, I needed to set the uncertainty this low to actually have it pass csynth. This has been an issue I've been looking at, but don't really know where that can be improved right now.

aehart replied:
Does changing the uncertainty actually change the timing estimate, i.e., the minimum clock period, or resource utilization that HLS reports, or does it just change whether it says it passes?

It sounds like it's the latter, in which case, you should comment out this line. We don't care too much about the timing estimate from HLS, since it's so unreliable and the estimate after implementation is presumably more accurate. And other modules, like the TE, fail timing according to HLS but pass according to the post-implementation estimate.

jasonfan393 replied:
yes, it's the latter, just commented it out.

aehart commented on project/script_TP.tcl:
```diff
@@ -11,24 +11,25 @@ open_project -reset trackletProcessor
 csynth_design
 cosim_design 
-export_design -format ip_catalog
+#export_design -format ip_catalog
 # Adding "-flow impl" runs full Vivado implementation, providing accurate resource use numbers (very slow).
-#export_design -format ip_catalog -flow impl
+export_design -format ip_catalog -flow impl
```
@jasonfan393 Please switch the comments on these lines so that the implementation is _not_ run by default. It's useful for us because it gives a more accurate timing estimate, but this Tcl file is used by the CI tests, which we don't want to take too long to run.

aehart reviewed (approved):
Looks good. Approved!

meisonlikesicecream commented:
Sorry for being a bit late to the party (and for writing it here; I don't think I have your email @jasonfan393), but this PR causes the VMRCM to crash for layer 1 due to lines like: https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/VMRouterCM.h#L191, as it tries to access an array element with index `-1`.

I could try and fix it myself in a new PR, but I'm not sure why we have this `(Layer) ? Layer-2 : Disk+4]` (why `-2` and `-4`?) or how the new `kNTEUnits` array work (why is it length 12? https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/Constants.h#L17). Maybe it's an easy fix for @jasonfan393? If not, let me know and I could try and figure out what's going on.

jasonfan393 commented:
> Sorry for being a bit late to the party (and for writing it here; I don't think I have your email @jasonfan393), but this PR causes the VMRCM to crash for layer 1 due to lines like: https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/VMRouterCM.h#L191, as it tries to access an array element with index `-1`.
> 
> I could try and fix it myself in a new PR, but I'm not sure why we have this `(Layer) ? Layer-2 : Disk+4]` (why `-2` and `-4`?) or how the new `kNTEUnits` array work (why is it length 12? https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/Constants.h#L17). Maybe it's an easy fix for @jasonfan393? If not, let me know and I could try and figure out what's going on.

Sorry, I probably should have left a comment here when I first made this change as it was just a quick fix to changing kNTEUnits for TP.  The new array just contains the value of NTEUnits for the corresponding seed, i.e. L1L2 => kNTEUnits[0], and the term in question "(Layer) ? Layer-2 : Disk+4] " is equal to layerdisk-1, but I think layerdisk isn't defined within VMRouterCM.h. The issue is just that since TP doesn't use a VMStubTEOuterMemoryCM in L1,  I hadn't thought of what to do in that case. I think I can come up with a quick fix for this, but I'm not super familiar with VMRCM so let me know if you think of any obvious fix here. 

Also, I can be reached at jf847@cornell.edu.
